packages/@aws-cdk/core/lib/asset-staging.ts

```diff
@@ -102,14 +113,27 @@ export class AssetStaging extends Construct {
       const bundlingStacks: string[] = this.node.tryGetContext(cxapi.BUNDLING_STACKS) ?? ['*'];
       const runBundling = bundlingStacks.includes(Stack.of(this).stackName) || bundlingStacks.includes('*');
       if (runBundling) {
-        // Determine the source hash in advance of bundling if the asset hash type
-        // is SOURCE so that the bundler can opt to re-use its previous output.
-        const sourceHash = hashType === AssetHashType.SOURCE
-          ? this.calculateHash(hashType, props.assetHash, props.bundling)
-          : undefined;
-
-        this.bundleDir = this.bundle(props.bundling, outdir, sourceHash);
-        this.assetHash = sourceHash ?? this.calculateHash(hashType, props.assetHash, props.bundling);
+        // Check if we already bundled this source path in this App (e.g. the same
+        // asset is used in multiple stacks). In this case we can completely skip
+        // file system and bundling operations.
+        this.cacheHash = crypto.createHash('sha256')
+          .update(path.resolve(this.sourcePath))
+          .update(JSON.stringify(props.bundling))
```

Comment:
Has `props.bundling` been normalized yet? It probably needs to be, otherwise 2 sets of properties that morally represent the same asset configuration render to 2 different keys (I'm thinking for example of `{ packaging }` specified or omitted because of the default)

Reply:
"Normalized" it by recursively sorting the keys after setting some defaults in the props.